### DISCUSSION TITLE
How to Migrate an Amazon MWS authorization to a Selling Partner API application

### DISCUSSION THREAD

**Discussion by Trimakas** [amzn/selling-partner-api-models#3121] [General] [answered]
[I've followed these instructions to a T.](https://github.com/amzn/selling-partner-api-docs/blob/main/guides/en-US/use-case-guides/authorization-api-use-case-guide/authorization-api-use-case-guide-v1.md#tutorial-migrate-an-amazon-mws-authorization-to-a-selling-partner-api-application)

And no luck. Has anyone actually converted MWS to SP-API without seller interaction as described in the docs. I've also opened a support ticket and this post on StackOverflow: [https://stackoverflow.com/questions/68868534/how-to-convert-amazon-mws-credentials-to-sp-api-creds](url)

**Comment by JordanForeman** [comment, marked as answer]
StackOverflow started a maintenance window just as I was typing my response, so I'll copy + paste my response here for now:

---

Unfortunately the specific Amazon docs that you link to don't tell the whole story. There are a few other requirements you'll need in order to get the `authorizationCode` response that you're looking for:

### Amazon OAuth Token

You'll need an access token from Amazon's OAuth API (an entirely different API). You can use [the `grantless` workflow](https://github.com/amzn/selling-partner-api-docs/blob/main/guides/en-US/developer-guide/SellingPartnerApiDeveloperGuide.md#grantless-operations) for this, since in your case the user hasn't actually authorized the SP-API yet:

```
POST https://api.amazon.com/auth/o2/token

body: {
    grant_type: 'client_credentials',
    scope: 'sellingpartnerapi::migration',
    client_id: 'amzn1.application-oa2-client.xxxxxxxxxxxxxxxxxxxxxxxx',
    client_secret: 'a473e76XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
}
```

This will return an `access_token` that you'll need for your actual migration request to `https://sellingpartnerapi-na.amazon.com/authorization/v1/authorizationCode`. The response will look something like:

```
{
    "access_token": "Atc|xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "scope": "sellingpartnerapi::migration",
    "token_type": "bearer",
    "expires_in": 3600
}
```

**Important:** Take the `access_token` value from that response and add it as an `x-amz-access-token` header to your `/authorization/v1/authorizationCode` request.

### Sign Your Request

This is the actual reason behind the error you're receiving. An unsigned request will not include the "authorization token" that you're being prompted for. 

You'll need to sign your request using Amazon's `SigV4` signing mechanism. It looks like you're using Ruby (HTTParty), so you can use the `aws-sdk`'s [Aws::Sigv4::Signer](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Sigv4/Signer.html) for this. You'll need to have setup IAM credentials as documented in [the generic developer guide](https://github.com/amzn/selling-partner-api-docs/blob/main/guides/en-US/developer-guide/SellingPartnerApiDeveloperGuide.md#creating-and-configuring-iam-policies-and-entities), and those credentials being provided to your `Aws::Sigv4::Signer` somehow (hardcoding, env vars, [Aws::SharedCredentials](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/SharedCredentials.html), etc.)

Request signing will result in a few proprietary headers being added to your request. Once this is done, you should have all that you need to make the request successfully.
